### PR TITLE
sql: introduce SHOW TENANT <name>

### PR DESCRIPTION
**Pull request by lidorcarmel**
This change adds:

`SHOW TENANT <name>`

To show the id, name and status of the tenant. The status can be ACTIVE/ADD/DROP.

The next step is to extend this to show replication status (more details
in https://github.com/cockroachdb/cockroach/issues/91261).

For example:
```
root@localhost:26257/defaultdb> show tenant my_tenant;
  id |   name    | status
-----+-----------+---------
   6 | my_tenant | ACTIVE
(1 row)


Time: 2ms total (execution 2ms / network 0ms)

root@localhost:26257/defaultdb> show tenant not_my_tenant;
ERROR: tenant "not_my_tenant" does not exist
SQLSTATE: 42704
```

Informs: https://github.com/cockroachdb/cockroach/issues/91261

Release note: None

Epic CRDB-18749

### CODE REVIEW

**Comment by lidorcarmel** [comment]
Please ignore for now.. just talked to Aditya, we don't really want the builtin so I'll change this pr to add the "SHOW TENANT" SQL.

**Comment by knz** [comment]
`show tenant status` plz. there will be more metadata in the future.

**Comment by knz** [comment]
if builtin fn, `tenant_status()` without `show` in the name.

**Comment by dt** [comment]
> show tenant status plz. there will be more metadata in the future.
@knz IIRC, the plan is `SHOW TENANT <name>` would have `status` one of the result columns and when there is more metadata in the future, we'd use additional result columns, rather than SHOW needing to have a separate SHOW command for each field of the tenant record

**Comment by knz** [comment]
thanks that workd

**Comment by dt** [comment]
should we add the ID and name while we're at it?

**Review by lidorcarmel — COMMENTED** [review]
so, I was thinking that long term users might not need to know the id? we can always query that directly but maybe it's too naive and we will need it often. how about leaving it out and adding it later if needed? I don't feel strongly.

and regarding the name, I can add. BTW, do we plan to support "SHOW TENANTS"?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-:-NH0z7cuAKCoO6Wnm4zv:b-x9ftp3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @knz)

**Comment by dt** [comment]
While I agree that in the long (or not even that long I hope) term, users should not need to know the ID, in in the short term it seems handy, as the ID is going to appear in places like key span bounds, and right now also in connection strings, certs, etc. My personal hope is that we fix those ASAP (and I think we actually can in a non-breaking way if we just treat ID == name for existing tenants) but as long as there are some APIs speaking in IDs, it seems handy to be able to know the ID for a given (named) tenant. 

I don't feel strongly either way though.

**Review by lidorcarmel — COMMENTED** [review]
makes sense to me, thanks, added!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-:-NH1WQdNAyXSjjakO_wS:b-uihzer)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @knz)

File: pkg/sql/tenant.go [hunk 1]
```diff
@@ -698,3 +698,33 @@ WHERE id = $1`, tenantID, tenantName); err != nil {
 
 	return nil
 }
+
+// GetTenantInfo implements the tree.TenantOperator interface.
+func (p *planner) GetTenantInfo(
+	ctx context.Context, tenantName roachpb.TenantName,
+) (*descpb.TenantInfo, error) {
```
Comment (stevendanna) [inline pkg/sql/tenant.go:705]: [nit] It's not particularly a lot of code so I don't mind the duplication too much, but this function is pretty much GetTenantRecordByName with an additional role check.  Perhaps we can call the other function?

File: pkg/sql/opaque.go [hunk 2]
```diff
@@ -252,6 +252,8 @@ func planOpaque(ctx context.Context, p *planner, stmt tree.Statement) (planNode,
 		return p.ShowHistogram(ctx, n)
 	case *tree.ShowTableStats:
 		return p.ShowTableStats(ctx, n)
+	case *tree.ShowTenant:
+		return p.ShowTenantNode(ctx, n)
```
Comment (stevendanna) [inline pkg/sql/opaque.go:256]: [nit] Most of the functions here don't have this `Node` suffix.

**Review by lidorcarmel — COMMENTED** [review]
btw this pr is ready for review.. the only open issue is whether we should SKIP DOC, discussed in https://github.com/cockroachdb/cockroach/pull/91983#discussion_r1024210259.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-:-NH1XI8-BpNVf3GHvMy5:b75mpmf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @knz)

File: pkg/sql/tenant.go [hunk 3]
```diff
@@ -698,3 +698,33 @@ WHERE id = $1`, tenantID, tenantName); err != nil {
...
+	)
+	if err != nil {
+		return nil, err
+	} else if row == nil {
+		return nil, pgerror.Newf(pgcode.UndefinedObject, "tenant %q does not exist", tenantName)
+	}
+
+	info := &descpb.TenantInfo{}
+	infoBytes := []byte(tree.MustBeDBytes(row[0]))
+	if err = protoutil.Unmarshal(infoBytes, info); err != nil {
```
Comment (stevendanna) [inline pkg/sql/tenant.go:726]: [nit you should probably just ignore and I probably shouldn't be typing] I think one of the reasons we use this `if err ...` construction is to restrict the scope of the `err` variable. But if we are assigning to an existing err variable, that kinda defeats the point.

**Review by stevendanna — CHANGES_REQUESTED** [review]
Nice. Left a few nitpics but it looks reasonable to me.

**Review by lidorcarmel — COMMENTED** [review]
thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-:-NH1xK-u9-b2Nuefl34w:bd5l6mr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @stevendanna)

___
*[`pkg/sql/opaque.go` line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-NH0ljSdxfz1Pbea8QLG-r2-256:-NH1tOG86dPNWZtFMydd:b6mp1m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6fb2bb9578a4cd486a5c3edb6a5f4200e844ebe/pkg/sql/opaque.go#L256)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] Most of the functions here don't have this `Node` suffix.
</blockquote></details>

indeed. done.
___
*[`pkg/sql/tenant.go` line 705 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-NGwmtdBlqyi24WtQ3aI-r2-705:-NH1tUSLDJhWecc4pjpK:b-bno2qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6fb2bb9578a4cd486a5c3edb6a5f4200e844ebe/pkg/sql/tenant.go#L705)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] It's not particularly a lot of code so I don't mind the duplication too much, but this function is pretty much GetTenantRecordByName with an additional role check.  Perhaps we can call the other function?
</blockquote></details>

done! I thought the pr that added GetTenantRecordByName was still inflight, but it's in, so rebased and fixed.
___
*[`pkg/sql/tenant.go` line 726 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-NGwmtdBlqyi24WtQ3aI-r2-726:-NH1wSpz6RYGS_iUm1cd:b-m0au0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6fb2bb9578a4cd486a5c3edb6a5f4200e844ebe/pkg/sql/tenant.go#L726)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit you should probably just ignore and I probably shouldn't be typing] I think one of the reasons we use this `if err ...` construction is to restrict the scope of the `err` variable. But if we are assigning to an existing err variable, that kinda defeats the point.
</blockquote></details>

oh no.. please feel free :) these lines are dropped now, but that makes sense to me!

**Review by lidorcarmel — COMMENTED** [review]
TFTR!
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91939#-:-NH5zV6XCa6vyLnrxkno:bihs03x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @stevendanna)

**Review by stevendanna — APPROVED** [review]
LGTM! May want to hold off on merging until after #92076 which will fix the string-ification in create and drop.

**Comment by lidorcarmel** [comment]
thanks! the other pr is on.
added a tiny test case to SHOW TENANT for a dropped tenant, and fixed a test with SHOW TENANT with quotes.

bors r+

**Pull request merged**